### PR TITLE
Cy an delete item

### LINKED ISSUE #11: 11. As a user, I want to be able to delete items from my shopping list so that my list isn’t cluttered with items I don’t want to buy in the future.
It’s possible our users may make an error when entering an item or may decide they won’t be buying a certain item again in the future. Let’s help them keep their lists tidy!

AC:

* User is able to delete an item from the shopping list
* Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
* Deletion should cause the associated record(s) in the database to be deleted

Notes:

* As noted in the wireframe https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png activating the "Delete" link should pop-up a confirmation dialog because this is a "destructive" action.
* Many teams jump to a modal as the solution here, but modals are fraught with accessibility concerns, so consider using a simple JavaScript confirm dialog.

### PR DESCRIPTION
**Pull request by sponberry**
## Description

Currently the app allows a user to add items, but not delete them. This PR covers a change to add a delete link for each item in the shopping list that will remove it from the UI and from the DB.

## Related Issue

closes #11 

## Acceptance Criteria

- [x]  User is able to delete an item from the shopping list 
- [x] Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
- [x] Deletion should cause the associated record(s) in the database to be deleted

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### After

delete_Item

## Testing Steps / QA Criteria

- From your terminal, checkout this branch with `git checkout cy-an-delete-item` and pull down with `git pull`
- Open the app with `npm start`
- Create a new list if you do not already have one. Add at least one item to your list if it's empty.
- Alongside the item(s) on your list, you should see a delete link.
- Click the delete link, a confirm dialog box should pop up. Select the cancel option - the dialog box should disappear and the item should remain on the list.
- Click the delete link again, this time select the OK option. The item should be deleted from the list and a message will appear in green above the list confirming item has been deleted for 10 seconds.

### CODE REVIEW

File: src/components/Item.jsx
Comment (nadinda) [inline src/components/Item.jsx]: is this meant to be here?

**Review by nadinda — APPROVED** [review]

File: src/components/ShowList.css
Comment (sponberry) [inline src/components/ShowList.css:41]: I've just reviewed the other pairs code and they've now created an Item.css file - perhaps we should consider moving all Item specific classes into an Item.css file now, so that our code aligns when merged ☺️

File: src/components/Item.jsx
Comment (sponberry) [inline src/components/Item.jsx:75]: This variable name has changed in the other pairs code, so we should probably change it here before we merge to (it's not checkedClassName I think) 😊

File: src/components/ShowList.css
Comment (kevinfalting) [reply src/components/ShowList.css:41]: I'd recommend waiting until one of the PRs are merged first, then merge main into the remaining one and resolve any conflicts there. 👍

File: src/components/Item.jsx
Comment (kevinfalting) [inline src/components/Item.jsx]: nit ❤️ : What would this look like if it was `if (!confirmation)`? You could have the function return early and keep the happy path outside of the `if` block.

```js
if (!confirmation) {
  return;
}

// continue with the delete...

File: src/components/Item.jsx
Comment (kevinfalting) [inline src/components/Item.jsx]: This would be a good opportunity to use async/await, could wrap it in a try/catch to handle the thrown error.

File: src/components/Item.jsx
Comment (kevinfalting) [inline src/components/Item.jsx]: Can we show an error message to the user if the delete fails? In this case, we're logging the error to the console, but not letting the user know something went wrong. They will only see that it looked like it went through, but the item remains.

File: src/components/ShowList.jsx
Comment (kevinfalting) [inline src/components/ShowList.jsx:65]: If we are deleting the very last item, this message won't show up since it depends on there being items in the list. We should be able to display the message regardless if there are items or not.

File: src/components/ShowList.jsx
Comment (kevinfalting) [inline src/components/ShowList.jsx:78]: Instead of passing the `setActionMessage` function from the useState hook, consider passing it a function that accepts a string and can handle the timeout. Right now, where this is being called, the caller also has to remember to remove the action message.

```js
const displayMessage = (message) => {
  setActionMessage(message)
  setTimeout(() => {
    setActionMessage(null);
  }, 10000);
}
```

This will keep all the logic together so the caller doesn't have to remember to also unset the action message.

Just pass it in here and the one below as:
```jsx
setActionMessage={displayMessage}
```

File: src/components/Item.jsx
Comment (kevinfalting) [inline src/components/Item.jsx]: 10 seconds feels just a tad too long to me. 😝

File: src/components/Item.jsx
Comment (kevinfalting) [inline src/components/Item.jsx]: What is this `href="x"` doing?

**Review by Amaka202 — APPROVED** [review]
Good job team. The feature works very well🎉

File: src/components/Item.jsx
Comment (Lukabaramishvili) [reply src/components/Item.jsx]: This is a great suggestion 🚀

File: src/components/Item.jsx
Comment (Lukabaramishvili) [inline src/components/Item.jsx]: I'm curious, any particular reason why are we using an anchor tag (`<a>`) here instead of a `<button>`? 

I found [this article](https://formidable.com/blog/2014/05/08/anchors-buttons-and-accessibility/) pretty interesting in terms of understanding which HTML tag would be semantically right to use as well as making it more accessible.

File: src/components/Item.jsx
Comment (Lukabaramishvili) [inline src/components/Item.jsx:77]: Currently, when the checkmark is checked, the item, as well as the delete button/text has a line-through on it. 
I think we don't need this style for the `delete` button. 

Screen Shot 2021-08-20 at 6 45 30 PM

The reason why it happens is that the delete button lives inside the `<li>` tag.
We could resolve it in two ways. Either don't extend this styling to the next element that lives inside `<li>` using CSS. Or style only the `<div>` (line 83) element with the line-through that renders the item name. 🙂

**Review by Lukabaramishvili — APPROVED** [review]
Beautiful implementation team 👏 Everything works as expected. You've done wonderful work. 

In addition to Kevin's amazing questions and suggestions, I left a couple of non-blocking ones too. 👍 🙌

File: src/components/Item.jsx
Comment (blackbureau) [reply src/components/Item.jsx:77]: Hi Luka, Thanks for these suggestions. I turned the delete 'link' into a button  as per your message below, which sorted out the strike-thru problem

File: src/components/Item.jsx
Comment (blackbureau) [reply src/components/Item.jsx]: Hi Kevin href doesn't seem to work without content (and initially we thought a link would be best), but has been removed now.

**Review by ishimwezachee — APPROVED** [review]

**Review by blackbureau — APPROVED** [review]

File: src/components/Item.jsx
Comment (sponberry) [reply src/components/Item.jsx]: It is not 😅 thanks Nadinda! ☺️

**Pull request merged**
